opennms-provision/opennms-detector-common/src/main/java/org/opennms/netmgt/provision/detector/common/DetectorRequestExecutorImpl.java

```diff
@@ -0,0 +1,168 @@
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2016 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2016 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.provision.detector.common;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import org.opennms.netmgt.provision.AsyncServiceDetector;
+import org.opennms.netmgt.provision.DetectFuture;
+import org.opennms.netmgt.provision.ServiceDetector;
+import org.opennms.netmgt.provision.SyncServiceDetector;
+import org.opennms.netmgt.provision.detector.registry.api.ServiceDetectorRegistry;
+
+public class DetectorRequestExecutorImpl implements DetectorRequestExecutor {
+
+    private Executor executor = Executors.newSingleThreadExecutor();
+
+    private ServiceDetectorRegistry serviceDetectorRegistry;
+
+    @Override
+    public CompletableFuture<DetectorResponseDTO> execute(
+            DetectorRequestDTO request) {
+
+        String serviceName = request.getServiceName();
+        String address = request.getAddress();
+        List<String> properties = request.getProperties();
+        Map<String, String> attributes = parse(properties);
+
+        ServiceDetector detector = serviceDetectorRegistry.getDetectorByServiceName(serviceName,
+                                                                                    attributes);
+
+        InetAddress ipAddress = null;
+        try {
+            ipAddress = InetAddress.getByName(address);
+        } catch (UnknownHostException e1) {
+            // TODO Auto-generated catch block
+            e1.printStackTrace();
```

Comment:
We should return a CompletableFuture and pass the exception using completeExceptionally() in this case.